src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs

```diff
@@ -0,0 +1,392 @@
+  {
+    private NUnitTestCaseBuilder _builder = new NUnitTestCaseBuilder();
+
+    #region Constructors
+
+    /// <summary>
+    /// Constructs the object responsible for obtaining test data.
+    /// </summary>
+    /// <param name="dataType">The type of data provider</param>
+    /// <param name="sourceName">The connection string for the data source. It could be a filename or DB connection string.</param>
+    /// <param name="rowCount">The number of rows to collect from the data source. By default it is all rows</param>
+    public DataSourceAttribute(DataSourceType dataType, string sourceName, int rowCount = (int)DataRowCount.All)
+    {
+      this.SourceName = sourceName;
+      this.DataType = dataType;
+      this.DataRows = rowCount;
+    }
+    
+    #endregion
+
+    #region Properties
+
+    /// <summary>a
+    /// The name of a file or connection string which contains the data
+    /// </summary>
+    public string SourceName { get; private set; }
+
+    /// <summary>
+    /// The data input type
+    /// </summary>
+    public DataSourceType DataType {get; private set;}
+
+    /// <summary>
+    /// The number of rows to read from the source. Zero takes a special meaning to be all rows. This is the default.
+    /// </summary>
+    public int DataRows { get; set;}
+    
+    #endregion
+    
+    #region ITestCaseSource Members
+    
+    /// <summary>
+    /// Returns a set of test case parameters for use as arguments
+    /// to a parameterized test method.
+    /// </summary>
+    /// <param name="method">The method for which data is needed.</param>
+    /// <returns></returns>
+    private IEnumerable<TestCaseParameters> GetTestCasesFor(IMethodInfo method)
+    {
+      List<TestCaseParameters> data = new List<TestCaseParameters>();
+
+      try
+      {
+        IEnumerable source = GetDataFromSource(method);
+
+        if(source != null)
+        {
+          #if NETCF
+          ParameterInfo[] parameters = method.IsGenericMethodDefinition ? new ParameterInfo[0] : method.GetParameters();
+          #else
+          IParameterInfo[] parameters = method.GetParameters();
+          #endif
+
+          foreach (object item in source)
+          {
+            var parms = item as TestCaseParameters;
+
+            if (parms == null)
+            {
+              object[] args = item as object[];
+              if (args != null)
+              {
+                #if NETCF
+                if (method.IsGenericMethodDefinition)
+                {
+                  var mi = method.MakeGenericMethodEx(args);
+                  parameters = mi == null ? new ParameterInfo[0] : mi.GetParameters();
+                }
+                #endif
+                if (args.Length != parameters.Length)
+                  args = new object[] { item };
+              }
+              else if (item is Array)
+              {
+                Array array = item as Array;
+
+                #if NETCF
```

Comment:
Another minor formatting issue, we put the #if statements in column 0 so they are obvious.